src/services/geometry/richgeo/IrtGeoPFRICH.h

```diff
@@ -14,9 +14,17 @@ namespace richgeo {
         IrtGeo("PFRICH",compactFile_,log_) { DD4hep_to_IRT(); }
       IrtGeoPFRICH(dd4hep::Detector *det_, std::shared_ptr<spdlog::logger> log_) :
         IrtGeo("PFRICH",det_,log_) { DD4hep_to_IRT(); }
-      ~IrtGeoPFRICH() {}
+      ~IrtGeoPFRICH();
 
     protected:
       void DD4hep_to_IRT() override;
+
+    private:
+      // FIXME: should be smart pointers, but even using unique/shared_ptr::get() causes issues in IRT...
```

Comment:
`IRT` class objects may sometimes assume ownership of these raw pointers, copy them, move them etc.. I haven't tracked down precisely what the issue is, but let me see if at least some of these can be made smart.

Reply:
These are the only ones that still work as a smart pointer:
```cpp
      std::shared_ptr<FlatSurface>             m_surfEntrance;
      std::shared_ptr<CherenkovPhotonDetector> m_irtPhotonDetector;
      std::shared_ptr<FlatSurface>             m_aerogelFlatSurface;
      std::shared_ptr<FlatSurface>             m_filterFlatSurface;
```
Converting the others to smart pointers breaks ROOT streamability; here is a stacktrace from [drich-dev auxfile creation](https://github.com/eic/drich-dev/blob/main/src/create_irt_auxfile.cpp), which uses these `IrtGeo`-derived classes to stream objects to ROOT files (not needed for EICrecon, but some other dRICH devs prefer to have this feature):
```
===========================================================
There was a crash.
This is the entire stack trace of all threads:
===========================================================
#0  0x00007f53f471e9d3 in wait4 () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007f53f469719b in ?? () from /lib/x86_64-linux-gnu/libc.so.6
#2  0x00007f53f539394d in TUnixSystem::StackTrace() () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libCore.so.6.26
#3  0x00007f53f53932e4 in TUnixSystem::DispatchSignals(ESignals) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libCore.so.6.26
#4  <signal handler called>
#5  0x00007f53f4a57d10 in vtable for __cxxabiv1::__vmi_class_type_info () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007f53f533031a in TClass::Store(TBuffer&) const () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libCore.so.6.26
#7  0x00007f53f4d92a31 in TBufferFile::WriteClass(TClass const*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#8  0x00007f53f4d936d1 in TBufferFile::WriteObjectClass(void const*, TClass const*, bool) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#9  0x00007f53f4d9affb in TBufferIO::WriteObjectAny(void const*, TClass const*, bool) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#10 0x00007f53f4d8e9ec in TBufferFile::WriteFastArray(void**, TClass const*, int, bool, TMemberStreamer*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#11 0x00007f53f4fff0be in int TStreamerInfo::WriteBufferAux<char**>(TBuffer&, char** const&, TStreamerInfo::TCompInfo* const*, int, int, int, int, int) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#12 0x00007f53f4e6123d in TStreamerInfoActions::GenericWriteAction(TBuffer&, void*, TStreamerInfoActions::TConfiguration const*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#13 0x00007f53f4d945b6 in TBufferFile::WriteClassBuffer(TClass const*, void*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#14 0x00007f53f428261d in OpticalBoundary::Streamer (this=0x564ef2fe03e0, R__b=...) at /home/dilks/d/drich-dev/irt/build/G__IRT.cxx:2340
#15 0x00007f53f4d9370f in TBufferFile::WriteObjectClass(void const*, TClass const*, bool) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#16 0x00007f53f4d9aeeb in TBufferIO::WriteObjectAny(void const*, TClass const*, bool) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#17 0x00007f53f4e13711 in TGenCollectionStreamer::WriteObjects(int, TBuffer&) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#18 0x00007f53f4de27f3 in TCollectionStreamer::Streamer(TBuffer&, void*, int, TClass*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#19 0x00007f53f4d8e375 in TBufferFile::WriteFastArray(void*, TClass const*, int, TMemberStreamer*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#20 0x00007f53f500189a in int TStreamerInfo::WriteBufferAux<char**>(TBuffer&, char** const&, TStreamerInfo::TCompInfo* const*, int, int, int, int, int) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#21 0x00007f53f4e6123d in TStreamerInfoActions::GenericWriteAction(TBuffer&, void*, TStreamerInfoActions::TConfiguration const*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#22 0x00007f53f4d945b6 in TBufferFile::WriteClassBuffer(TClass const*, void*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#23 0x00007f53f4282e5d in CherenkovPhotonDetector::Streamer (this=0x564ef2fc8f00, R__b=...) at /home/dilks/d/drich-dev/irt/build/G__IRT.cxx:2406
#24 0x00007f53f4d9370f in TBufferFile::WriteObjectClass(void const*, TClass const*, bool) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#25 0x00007f53f4d9aeeb in TBufferIO::WriteObjectAny(void const*, TClass const*, bool) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#26 0x00007f53f4e13711 in TGenCollectionStreamer::WriteObjects(int, TBuffer&) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#27 0x00007f53f4de27f3 in TCollectionStreamer::Streamer(TBuffer&, void*, int, TClass*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#28 0x00007f53f4d8e375 in TBufferFile::WriteFastArray(void*, TClass const*, int, TMemberStreamer*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#29 0x00007f53f500189a in int TStreamerInfo::WriteBufferAux<char**>(TBuffer&, char** const&, TStreamerInfo::TCompInfo* const*, int, int, int, int, int) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#30 0x00007f53f4e6123d in TStreamerInfoActions::GenericWriteAction(TBuffer&, void*, TStreamerInfoActions::TConfiguration const*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#31 0x00007f53f4d945b6 in TBufferFile::WriteClassBuffer(TClass const*, void*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#32 0x00007f53f4285209 in CherenkovDetector::Streamer (this=0x564ed4553950, R__b=...) at /home/dilks/d/drich-dev/irt/build/G__IRT.cxx:2696
#33 0x00007f53f4d9370f in TBufferFile::WriteObjectClass(void const*, TClass const*, bool) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#34 0x00007f53f4d9aeeb in TBufferIO::WriteObjectAny(void const*, TClass const*, bool) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#35 0x00007f53f4d8e9ec in TBufferFile::WriteFastArray(void**, TClass const*, int, bool, TMemberStreamer*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#36 0x00007f53f4ff8922 in int TStreamerInfo::WriteBufferAux<TVirtualCollectionProxy>(TBuffer&, TVirtualCollectionProxy const&, TStreamerInfo::TCompInfo* const*, int, int, int, int, int) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#37 0x00007f53f4e94d0b in TStreamerInfoActions::GenericLooper::GenericWrite(TBuffer&, void*, void const*, TStreamerInfoActions::TLoopConfiguration const*, TStreamerInfoActions::TConfiguration const*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#38 0x00007f53f4d8d20c in TBufferFile::ApplySequence(TStreamerInfoActions::TActionSequence const&, void*, void*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#39 0x00007f53f5001a5c in int TStreamerInfo::WriteBufferAux<char**>(TBuffer&, char** const&, TStreamerInfo::TCompInfo* const*, int, int, int, int, int) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#40 0x00007f53f4e6123d in TStreamerInfoActions::GenericWriteAction(TBuffer&, void*, TStreamerInfoActions::TConfiguration const*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#41 0x00007f53f4d945b6 in TBufferFile::WriteClassBuffer(TClass const*, void*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#42 0x00007f53f428565b in CherenkovDetectorCollection::Streamer (this=0x564ed470c7a0, R__b=...) at /home/dilks/d/drich-dev/irt/build/G__IRT.cxx:2729
#43 0x00007f53f4e3a3f6 in TKey::TKey(TObject const*, char const*, int, TDirectory*) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#44 0x00007f53f4dfa245 in TFile::CreateKey(TDirectory*, TObject const*, char const*, int) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#45 0x00007f53f4deb66e in TDirectoryFile::WriteTObject(TObject const*, char const*, char const*, int) () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libRIO.so.6.26
#46 0x00007f53f5275649 in TObject::Write(char const*, int, int) const () from /opt/software/linux-debian-x86_64_v2/gcc-12.2.0/root-6.26.10-ypxsyrtxgzrojuy7ainximgo4er5zmmz/lib/root/libCore.so.6.26
#47 0x0000564ece45165a in main (argc=2, argv=0x7fff5edb6d18) at src/create_irt_auxfile.cpp:63
===========================================================
```

Reply:
Do you know of a way to have ROOT I/O play nicely with streaming objects owned by smart pointers?

Reply:
New comment is more specific, thanks.

> `IRT` class objects may sometimes assume ownership of these raw pointers, copy them, move them etc.. I haven't tracked down precisely what the issue is, but let me see if at least some of these can be made smart.

But you still call `operator delete *` on the objects in the end, it should not make a difference if you do it twice with a smart pointer or manually.

Reply:
Based on comments from @alexander-kiselev below, and on private discussions we've had in the past, I think it is safer to just use these raw pointers here. I tried my best to use smart pointers everywhere in the algorithm too, `IrtCherenkovParticleID`, but again passing `std::shared_ptr<T>::get` to IRT methods tends to cause unexpected results or seg faults.

As long as we confirm that any `operator new *` has a corresponding `operator delete *`, we should be okay. There are [leaks in IRT, however](https://github.com/eic/irt/issues/36), where some destructors need implementation. Since we don't intend for this IRT to be run in production yet, these leaks should not block this PR's approval (but should be addressed ASAP).